src/version.rs

```diff
@@ -404,6 +515,143 @@ pub fn downloads(req: &mut Request) -> CargoResult<Response> {
     Ok(req.json(&R{ version_downloads: downloads }))
 }
 
+#[derive(Debug)]
+struct MaxOrNone<T>(Option<T>);
+
+impl<T> MaxOrNone<T>
+    where T: Ord,
+{
+    fn new() -> Self {
+        MaxOrNone(None)
+    }
+
+    fn push(&mut self, value: T) {
+        match self.0 {
+            Some(ref mut old) => {
+                if value.cmp(old) == ::std::cmp::Ordering::Greater {
+                    ::std::mem::replace(old, value);
+                }
+            }
+            None => self.0 = Some(value),
+        }
+    }
+}
+
+pub struct MaxBuildInfo {
+    pub stable: Option<semver::Version>,
+    pub beta: Option<time::Timespec>,
+    pub nightly: Option<time::Timespec>,
+}
+
+impl MaxBuildInfo {
+    pub fn encode(self) -> EncodableMaxBuildInfo {
+        EncodableMaxBuildInfo {
+            stable: self.stable.map(|v| v.to_string()),
+            beta: self.beta.map(::encode_time),
+            nightly: self.nightly.map(::encode_time),
+        }
+    }
+}
+
+#[derive(Debug, Default, RustcEncodable, RustcDecodable)]
+pub struct EncodableMaxBuildInfo {
+    pub stable: Option<String>,
+    pub beta: Option<String>,
+    pub nightly: Option<String>,
+}
+
+impl BuildInfo {
+    pub fn max<I>(build_infos: I) -> CargoResult<MaxBuildInfo>
+        where I: IntoIterator<Item = BuildInfo>
+    {
+        let mut stable = MaxOrNone::new();
+        let mut beta = MaxOrNone::new();
+        let mut nightly = MaxOrNone::new();
+
+        for row in build_infos {
+            let BuildInfo { rust_version, .. } = row;
+            let rust_version: ChannelVersion = rust_version.parse()?;
+
+            match rust_version {
+                ChannelVersion::Stable(semver) => stable.push(semver),
+                ChannelVersion::Beta(date) => beta.push(date),
+                ChannelVersion::Nightly(date) => nightly.push(date),
+            }
+        }
```

Comment:
>  just to avoid the one allocation of an intermediate vec

To be honest, I've never really worked on mission-critical infrastructure supporting the entire ecosystem behind a language, so I don't really know when it's appropriate. From the comment, I assume you think it's not, so I'll defer to your greater experience.